Sources/FoundationEssentials/FileManager/FileOperations.swift

```diff
@@ -342,52 +348,69 @@ enum _FileOperations {
-            guard filemanager?._shouldRemoveItemAtPath(directory) ?? true else { continue }
-            try directory.withNTPathRepresentation {
-                if RemoveDirectoryW($0) { return }
-                let dwError: DWORD = GetLastError()
-                guard dwError == ERROR_DIR_NOT_EMPTY else {
-                    let error = CocoaError.removeFileError(dwError, directory)
-                    guard (filemanager?._shouldProceedAfter(error: error, removingItemAtPath: directory) ?? false) else {
-                        throw error
+            var stack = [path]
+            while let directory = stack.popLast() {
+                try directory.withNTPathRepresentation {
+                    let path = String(decodingCString: $0, as: UTF16.self)
+
+                    guard filemanager?._shouldRemoveItemAtPath(path) ?? true else { return }
+
+                    if RemoveDirectoryW($0) { return }
+                    var dwError: DWORD = GetLastError()
+                    guard dwError == ERROR_DIR_NOT_EMPTY else {
+                        let error = CocoaError.removeFileError(dwError, directory)
+                        guard (filemanager?._shouldProceedAfter(error: error, removingItemAtPath: path) ?? false) else {
+                            throw error
+                        }
+                        return
                     }
-                    return
-                }
-                stack.append(directory)
 
-                for entry in _Win32DirectoryContentsSequence(path: directory, appendSlashForDirectory: false, prefix: [directory]) {
-                    try entry.fileNameWithPrefix.withNTPathRepresentation {
-                        if entry.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY == FILE_ATTRIBUTE_DIRECTORY,
-                                entry.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT != FILE_ATTRIBUTE_REPARSE_POINT {
-                            stack.append(entry.fileNameWithPrefix)
-                        } else {
-                            if entry.dwFileAttributes & FILE_ATTRIBUTE_READONLY == FILE_ATTRIBUTE_READONLY {
-                                guard SetFileAttributesW($0, entry.dwFileAttributes & ~FILE_ATTRIBUTE_READONLY) else {
-                                    throw CocoaError.removeFileError(GetLastError(), path)
+                    for entry in _Win32DirectoryContentsSequence(path: directory, appendSlashForDirectory: false, prefix: [directory]) {
+                        try entry.fileNameWithPrefix.withNTPathRepresentation {
+                            let path = String(decodingCString: $0, as: UTF16.self)
+
+                            if entry.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY == FILE_ATTRIBUTE_DIRECTORY,
+                                    entry.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT != FILE_ATTRIBUTE_REPARSE_POINT {
+                                if filemanager?._shouldRemoveItemAtPath(path) ?? true {
```

Comment:
I think we check this whenever we pop something off of the stack which means we'd ask about this directory twice. Is that intentional?

Reply:
Right, that was part of the problem. We now will do a clean up post-traversal which ensures that the question is only asked once. Fortunately, the test coverage ensures that we get this correct.

Reply:
I'm not certain the unit test actually tests this case, something like:

- foo (DIR)
    - bar (DIR)
        - baz (FILE)
 
In that case, I think we:

- Pop the first thing off the stack (foo) and ask if it should be removed
- Since it's a directory we enumerate its contents
    - While doing this, we see bar is a directory and ask if it should be removed, then put it on the stack
- Then we hit the remove call at the end which fails because foo isn't empty
- Then we pop the next thing off the stack (bar) and ask if it should be removed (a second time)
- It's not empty so we iterate it's contents
    - When we see baz we see it's a file and immediately ask and remove it
- Then we fall down to the remove call at the end which removes bar because it is now empty

I think that leaves us with not removing foo and asking about bar twice right? I think the case that would hit this is removing a directory with a nested directory specifically

Reply:
Added that as a test case and tweaked the logic to avoid the duplicate check.